### ISSUE TITLE
Question - v4.1 

### ISSUE THREAD

**Issue by codeclout** [samvera/serverless-iiif#88]
Hi,

Thanks again for the new layer that support JP2000... I just had time allocated to spin it up this sprint. I'm getting the following error. 
```json
"errorMessage": "\nSomething went wrong installing the \"sharp\" module\n\nCannot find module '../build/Release/sharp-linux-x64.node'\nRequire stack:\n- /var/task/node_modules/sharp/lib/sharp.js\n- /var/task/node_modules/sharp/lib/constructor.js\n- /var/task/node_modules/sharp/lib/index.js\n- /var/task/node_modules/iiif-processor/lib/transform.js\n- /var/task/node_modules/iiif-processor/index.js\n- /var/task/index.js\n- /var/runtime/index.mjs\n\nPossible solutions:\n- Install with verbose logging and look for errors: \"npm install --ignore-scripts=false --foreground-scripts --verbose sharp\"\n- Install for the current linux-x64 runtime: \"npm install --platform=linux --arch=x64 sharp\"\n- Consult the installation documentation: https://sharp.pixelplumbing.com/install"
```
Any thoughts as to what we're doing incorrectly? 

Thanks in advance

**Comment by mbklein** [comment]
What process are you using to spin it up? It looks like you might have a `node_modules` directory under `build` that is built for the wrong architecture, so `npm` thinks sharp is already installed when it actually needs to be rebuilt for linux x64. But without seeing the steps involved, I can't be sure.

**Comment by codeclout** [comment]
@mbklein  - thank you!

I made the following edits as a first run of http source support from a forked version of this repo:
- https://github.com/codeclout/serverless-iiif/commit/cf2feb82872da16547309bed468c4dd985cb07e7

Also pointed the `src` and `dependency` directories to zip files that were uploaded to S3.... due to errors being generated from the relative path
- https://github.com/codeclout/serverless-iiif/commit/fd9e580d743888212d0b7897705d370e13dce4cc

Lastly, used the `standalone template`, edited in the first commit in CloudFormation with all of the necessary params.

Thanks in Advance,
Brian

**Comment by mbklein** [comment]
Are you trying to publish this to SAR and deploy from there? Or deploying using the command line instructions?

To be honest, I'm not completely familiar with how the command line deploy script works since @orangewolf's edits. I'd recommend reverting https://github.com/codeclout/serverless-iiif/commit/fd9e580d743888212d0b7897705d370e13dce4cc, then running the following from the `sam/standalone` directory:

```
$ rm -rf .aws-sam # to remove any existing build artifacts
$ sam build --use-container
$ sam validate
$ sam deploy --guided
```

When it prompts you for the `SharpLayer` variable, leave it blank (default) so it'll use the JP2-compatible layer. What region are you deploying in? I'll make sure there's a JP2 layer for that region. (There should be – I think I deployed it to all the North American regions.)

**Comment by codeclout** [comment]
@mbklein  - Thanks again!

I was trying to deploy from S3... ultimately, if this POC is approved to move out of the sandbox we will need to automate the deployment... although I haven't given a ton of thought to that just yet. Publishing to SAR sounds like a good option. I have an ec2 with permissions to talk to `all the things` so I'll give the command line build with your recommendations a go.

We are in us-east-1 (GovCloud). 

Thanks again for all of your help... highly appreciated!

**Comment by mbklein** [comment]
I would recommend against publishing to SAR – it sounds like an extra step that won't be necessary for your use case. Depending on what you use for CI, automated deployments via the `sam` CLI can be pretty straightforward.

I don't think I have the jp2 sharp layer deployed to any of the govcloud regions but I can look into it. There's a good change I don't have access to them, though.

**Comment by codeclout** [comment]
@mbklein - thanks for all of your help.

The consensus from our broader team is that we would like to use your SAR template for deployment. Is there a way to point the SAR v4.1.1 to our forked repo for http source development efforts? 

Thanks in advance

**Comment by codeclout** [comment]
@mbklein 

Would a pull request allowing for the override of the code source via the SAM template be considered? It is a change our broader team needs, but would like to keep the ability to upgrade from SAR.

**Comment by mbklein** [comment]
I'd take a look at a PR for overriding the code source as long as the default behavior didn't change, and if it had some good documentation to go with it.

**Comment by codeclout** [comment]
@mbklein - I have been periodically working on a pull request that will address not only the code source location of the SAM template, but also HTTP as an alternative data source. I wanted to bring back a couple of decisions for which you may have a preference. 

- Do you have a preference as to where to retrieve the HTTP Source Host string from...? (e.g. - SAM template or Environment)

I was also evaluating EXIF processor modules from NPM that would allow for retrieving dimension metadata from the data returned from the HTTP Source. 

- I have been evaluating [https://github.com/hMatoba/piexifjs](https://github.com/hMatoba/piexifjs)... but was wondering if there was a preference?

Thanks in Advance,
Brian

**Comment by mbklein** [comment]
> Do you have a preference as to where to retrieve the HTTP Source Host string from...? (e.g. - SAM template or Environment)

If the HTTP source is a deploy-time configuration option, it should go in the SAM template. Then SAM can write it to the IIIF function's environment when it's deployed.

> I have been evaluating https://github.com/hMatoba/piexifjs... but was wondering if there was a preference?

We have struggled with various tools for reading EXIF data from images in lambdas. The best we've found so far is pretty cumbersome – a full installation of [exiftool](https://exiftool.org/) in a lambda layer which we call out to from nodejs in a subprocess.

That said, if it's just dimension info you're looking for, the existing [`probe-image-size`](https://www.npmjs.com/package/probe-image-size) package does exactly that, over HTTP. In fact, the default implementation uses a signed URL to allow it to access S3 objects over HTTPS.
